File: #127_Word_Ladder/Solution.java
import java.util.Deque;
import java.util.Set;
import java.util.LinkedList;
import java.util.HashSet;

public class Solution {
    public int ladderLength(String beginWord, String endWord, Set<String> wordDict) {
        if (beginWord.equals(endWord)) return 0;
        if (wordDict.isEmpty()) return 0;
        int len = beginWord.length();
        if (len == 1) {
        	if (wordDict.contains(beginWord) && wordDict.contains(endWord)) return 2;
        	return 0;
        }
        int level = 1;
        int prevLeveSize = 1;
        int nextLeveSize = 0;
        Deque<String> queue = new LinkedList<String>();
        Set<String> visited = new HashSet<String>();
        queue.add(beginWord);
        visited.add(beginWord);
        while(!queue.isEmpty()) {
        	prevLeveSize--;
        	String str = queue.remove();
        	char[] charOfStr = str.toCharArray();
        	for (int index = 0; index < len; index++) {
	        	char tmpChar = charOfStr[index];
        		for (char c = 'a'; c <= 'z'; c++) {
        			if (c == tmpChar) continue;
        			charOfStr[index] = c;
        			String tmp = new String(charOfStr);
        			if (tmp.equals(endWord)) {
        				return level + 1;
        			}
        			if (!visited.contains(tmp) && wordDict.contains(tmp)) {
    					nextLeveSize++;
        				visited.add(tmp);
        				queue.add(tmp);
        			}
        			charOfStr[index] = tmpChar;
        		}
        	}
        	if (prevLeveSize == 0) {
        		level++;
        		prevLeveSize = nextLeveSize;
        		nextLeveSize = 0;
        	}
        }
        return 0;
    }

    public static void main(String[] args) {
    	Solution sol = new Solution();
    	String start = "hit";
    	String end = "cog";
    	HashSet<String> dict = new HashSet<String>();
    	dict.add("hot");
    	dict.add("dot");
    	dict.add("dog");
    	dict.add("lot");
    	dict.add("log");
    	System.out.println(sol.ladderLength(start, end, dict));
    }
}

File: #127_Word_Ladder/Solution.py
'''
DFS solution ETL
'''

import sys

class Solution:
	# @param {string} beginWord
	# @param {string} endWord
	# @param {set<string>} wordDict
	# @return {integer}
	def ladderLength(self, beginWord, endWord, wordDict):
		if beginWord == endWord:
			return 1
		if not wordDict:
			return 0
		re = sys.maxint
		wordList = list(beginWord)
		for i in range(len(wordList)):
			for x in range(ord('a'), ord('z') + 1):
				if chr(x) != beginWord[i]:
					wordList[i] = chr(x)
					newWord = ''.join(wordList)
					if newWord == endWord:
						return 2
					if newWord in wordDict:
						wordDict.remove(newWord)
						ans = self.ladderLength(newWord, endWord, wordDict)
						if ans > 0:
							re = min(ans, re)
						wordDict.add(newWord)
					wordList[i] = beginWord[i]
		if re == sys.maxint:
			return 0
		return re + 1

sol = Solution()
dic = set(["hot","dot","dog","lot","log"])
print sol.ladderLength("hit", "cog", dic)